lib/managementClient.ts

```diff
@@ -0,0 +1,231 @@
+      debug(`[${connection.options.id}] Created sender "${this._mgmgtReqResLink.sender.name}" and receiver "${this._mgmgtReqResLink.receiver.name}" links for $management endpoint.`);
+    }
+  }
+
+  /**
+   * @private
+   * Helper method to make the management request
+   * @param {Connection} connection - The established amqp connection
+   * @param {string} type - The type of entity requested for. Valid values are "eventhub", "partition"
+   * @param {string | number} [partitionId] - The partitionId. Required only when type is "partition".
+   */
+  private async _makeManagementRequest(connection: any, type: "eventhub" | "partition", partitionId?: string | number): Promise<any> {
+    if (partitionId && typeof partitionId !== "string" && typeof partitionId !== "number") {
+      throw new Error("'partitionId' is a required parameter and must be of type: 'string' | 'number'.");
+    }
+    return new Promise(async (resolve: any, reject: any) => {
+      try {
+        const endpoint = Constants.management;
+        const replyTo = uuid();
+        const request: any = {
+          body: Buffer.from(JSON.stringify([])),
+          properties: {
+            message_id: uuid(),
+            reply_to: replyTo
+          },
+          application_properties: {
+            operation: Constants.readOperation,
+            name: this.entityPath as string,
+            type: `${Constants.vendorString}:${type}`
+          }
+        };
+        if (partitionId && type === Constants.partition) {
+          request.application_properties.partition = partitionId;
+        }
+        await defaultLock.acquire(this.managementLock, () => { return this._init(connection, endpoint, replyTo); });
+        // TODO: Handle timeout incase SB/EH does not send a response.
+        const messageCallback = ({ message, delivery }: any) => {
+          // remove the event listener as this will be registered next time when someone makes a request.
+          this._mgmgtReqResLink!.receiver.removeListener(Constants.message, messageCallback);
+          const code: number = message.application_properties[Constants.statusCode];
+          const desc: string = message.application_properties[Constants.statusDescription];
+          debug(`[${connection.options.id}] $management request: \n`, request);
+          debug(`[${connection.options.id}] $management response: \n`, message);
+          if (code === rheaPromise.AmqpResponseStatusCode.OK || code === rheaPromise.AmqpResponseStatusCode.Accepted) {
+            return resolve(message.body);
+          } else {
+            const condition = ConditionStatusMapper[code] || "amqp:internal-error";
+            let e: rheaPromise.AmqpError = {
+              condition: condition,
+              description: desc
+            };
+            return reject(translate(e));
+          }
+        };
+        this._mgmgtReqResLink!.receiver.on(Constants.message, messageCallback);
```

Comment:
when is it possible to not have a `this._mgmgtReqResLink`? (also i think there's a typo: mgmgt -> mgmt)

Reply:
(i'm making comments on these because i find that when there's a `!` assertion it usually means that it is actually possible for something to be null/undefined or there's a way to declare it as non-nullable from the start)

Reply:
I do not create the mgmtReqResLink in the constructor. It is created inside the init() method over [here](https://github.com/Azure/azure-event-hubs-node/pull/47/commits/d570fe94d77eee17e6ba2897a8cc1126313bc0d3#diff-1bc20a630438f30ef4378368cf293caeR165). TS frowns for not setting a value inside the constructor for a required property. This management client is created while instantiating the EH client. However the $management session is only created when someone asks for a hub runtime, partitionIds or partition runtime.